ariadne/asgi.py

```diff
@@ -0,0 +1,230 @@
+        self, request: Request
+    ) -> Tuple[str, Optional[dict], Optional[str]]:
+        if request.method == "GET":
+            return self.extract_data_from_request_query(request.query_params)
+        if request.headers.get("Content-Type") != DATA_TYPE_JSON:
+            raise HttpBadRequestError(
+                "Posted content must be of type {}".format(DATA_TYPE_JSON)
+            )
+
+        data = await request.json()
+        return self.extract_data_from_request_data(data)
+
+    async def extract_data_from_websocket(
+        self, message: dict
+    ) -> Tuple[str, Optional[dict], Optional[str]]:
+        payload = cast(dict, message.get("payload"))
+        if not isinstance(payload, dict):
+            raise GraphQLError("Payload must be an object")
+
+        query = cast(str, payload.get("query"))
+        variables = cast(dict, payload.get("variables"))
+        operation_name = cast(str, payload.get("operationName"))
+
+        return query, variables, operation_name
+
+    async def render_playground(self, request: Request) -> HTMLResponse:
+        return HTMLResponse(PLAYGROUND_HTML)
+
+    async def graphql_http_server(self, request: Request) -> Response:
+        try:
+            query, variables, operation_name = await self.extract_data_from_request(
+                request
+            )
+            document = parse(query)
+            result = await graphql(
+                self.schema,
+                query,
+                root_value=await self.root_value_for_document(document, variables),
+                context_value=await self.context_for_request(request),
+                variable_values=variables,
+                operation_name=operation_name,
+            )
+        except GraphQLError as error:
+            response = {"errors": [{"message": error.message}]}
+            return JSONResponse(response)
+        except HttpError as error:
+            response = error.message or error.status
+            return Response(response, status_code=400)
+        else:
+            response = {"data": result.data}
+            if result.errors:
+                response["errors"] = [format_error(e) for e in result.errors]
+            return JSONResponse(response)
+
+    async def observe_async_results(
+        self, results: AsyncGenerator, operation_id: str, websocket: WebSocket
+    ) -> None:
+        async for result in results:
+            payload = {}
+            if result.data:
+                payload["data"] = result.data
+            if result.errors:
+                payload["errors"] = [format_error(e) for e in result.errors]
+            await self.send_json(
+                websocket, {"type": GQL_DATA, "id": operation_id, "payload": payload}
+            )
+        await self.send_json(websocket, {"type": GQL_COMPLETE, "id": operation_id})
+
+    async def keep_connection_alive(self, websocket: WebSocket):
+        if not self.keepalive:
+            return
+        while True:
```

Comment:
```suggestion
        while websocket.application_state != WebSocketState.DISCONNECTED::
```

We can drop bottom `if/break` that way